rnn_speech_recognition/pytorch/README.md

```diff
@@ -1,44 +1,192 @@
+1. In the train pipeline, an adaptive specaugment augmentation is applied ([arxiv](https://arxiv.org/abs/1912.05533), [code](https://github.com/mwawrzos/training/blob/rnnt/rnn_speech_recognition/pytorch/common/data/features.py#L44-L117)). In the evaluation pipeline, this step is omitted;
+1. to reduce accelerator memory usage, frames are spliced (stacked three times, and subsampled three times) ([code](https://github.com/mwawrzos/training/blob/rnnt/rnn_speech_recognition/pytorch/common/data/features.py#L144-L165));
+
 ### Training and test data separation
-How is the test set extracted?
+Dataset authors separated it to test and training subsets. For this benchmark, training is done on train-clean-100, train-clean-360 and train-other-500 subsets. Evaluation is done on dev-clean subset.
+
 ### Training data order
-In what order is the training data traversed?
+To reduce data padding in minibatches, data bucketing is applied.
+The algorithm is implemented here:
+[link](https://github.com/mlcommons/training/blob/2126999a1ffff542064bb3208650a1e673920dcf/rnn_speech_recognition/pytorch/common/data/dali/sampler.py#L65-L105)
+and can be described as follows:
+1. drop samples longer than a given threshold;
+1. sort data by audio length;
+2. split data into 6 equally sized buckets;
+3. for every epochs:
+    1. shuffle data in each bucket;
+    2. as long as all samples are not divisible by global batch size, remove random element from random bucket;
+    3. concatenate all buckets;
+    4. split samples into minibatches;
+    5. shuffle minibatches in the epoch.
+
 ### Test data order
-In what order is the test data traversed?
-### Simulation environment (RL models only)
-Describe simulation environment briefly, if applicable. 
+Test data order is the same as in the dataset.
+
 # 4. Model
 ### Publication/Attribution
-Cite paper describing model plus any additional attribution requested by code authors 
+To the best of our knowledge, there is no single publication describing RNN-T training on LibriSpeech,
+or another publicly available dataset of reasonable size. For that reason, the reference will be a
+collection of solutions from several works. It is based on the following articles:
+* Graves 2012 - an invention of RNN-Transducer: https://arxiv.org/abs/1211.3711
+* Rao 2018 - time reduction in the acoustic model, internal dataset: https://arxiv.org/abs/1801.00841
+* Zhang 2020 - Transformer-transducer publication. It includes bi-directional LSTM RNN-T result on LibriSpeech: https://arxiv.org/abs/2002.02562
+* Park 2019 - adaptive spec augment, internal dataset: https://arxiv.org/abs/1912.05533
+* Guo 2020 - RNN-T trained with vanilla LSTM, internal dataset: https://arxiv.org/abs/2007.13802
+
 ### List of layers 
-Brief summary of structure of model
+Model structure is described in the following picture:
+![model layers structure](./rnnt_layers.svg "RNN-T model structure")
```

Comment:
One detail in the model structure is wrong. It says that it uses a `3x stack, 3x subsample` layer for the audio-encoding network. But it really is a `2x stack, 2x subsmple` layer as indicated here: [enc-stack-time-factor](https://github.com/mwawrzos/training/blob/rnnt/rnn_speech_recognition/pytorch/configs/baseline_v3-1023sp.yaml#L70)

Reply:
Good catch!

Reply:
fixed in 1c5d03a7f8a6a22ca4bf53c3f55ba5d61ab8b28e

Reply:
sorry to bug on this - but one more detail in the model structure is incorrect. The last FC in the Joint Net is `FC 512 -> 1024`  (not 1025). The number of tokenizer labels is 1023 and plus the blank symbol gives total number-of-symbols=1024.

Reply:
Thanks! Fixed in https://github.com/mlcommons/training/pull/430/commits/371f2b67518291f5aa4d7044a436832a9cbaf66e